0016-pre-shared-key/0016-pre-shared-key.md

```diff
@@ -0,0 +1,237 @@
+# Pre-Shared Key
+
+The Pre-Shared Key (PSK) protocol is an end-to-end transport protocol, used by the sender and receiver of an ILP payment to decide on a condition and fulfillment for a payment. The protocol also encrypts any additional data sent along with the payment, using [AES-256-CTR](https://en.wikipedia.org/wiki/Advanced_Encryption_Standard). The PSK data is authenticated through an [HMAC-SHA-256](https://en.wikipedia.org/wiki/Hash-based_message_authentication_code) which is used to generate the fulfillment of a PSK payment. Pseudocode for this protocol can be read [at the bottom of this spec](#pseudocode).
+
+As the name suggests, PSK relies on a pre-shared secret key. In application-layer protocols like SPSP, the receiver generates the shared key and shares it with the sender over an end-to-end secure communication channel such as HTTPS. Alternatively, a diffie-hellman key exchange could be used directly to generate the shared secret key.
+
+An advantage of PSK is that the pre-shared secret key only needs to be shared once. As long as both parties possess this key and are listening for transfers, they can send payments between one another.
+
+A disadvantage of PSK is that it is repudiable. Knowledge of a payment's fulfillment cannot be used as cryptographic proof that the receiver completed the payment, because both sender and receiver know how to generate the fulfillment with their pre-shared secret key. This causes no issues unless you have some other framework on top of ILP that relies on this cryptographic proof.
+
+## Flow
+
+1. The pre-shared secret key is shared out of band.
+2. The sender creates a random nonce, and calculates an HMAC of the nonce using the pre-shared secret key as the key. The result of this HMAC will be referred to as the payment encryption key.
+3. The sender constructs the PSK data:
+    1. The sender starts with the PSK status line: `PSK/1.0\n`
+    2. The sender appends the [public headers](#public-headers) (including the nonce), followed by `\n\n`.
+    3. The remainder of the PSK data after this point will be encrypted using AES-256-CTR with the payment encryption key. Note: The ciphertext is raw binary data, and is not base64 encoded.
+    4. The sender appends the [private headers](#private-headers), followed by `\n\n`.
```

Comment:
Wouldn't the `\n\n` be part of the encrypted ciphertext? It's a little confusing when written out this way.